### PR TITLE
Feature/test examples

### PR DESCRIPTION
**Pull request by MaGering**
Fix Issue https://github.com/greco-project/pvcompare/issues/228

With this PR tests are added which check if the examples in pvcompare run.

Running with `exit code == 0` leads to:

```
os.system("python example_to_be_tested.py") == 0
```

The following steps were realized, as well (required):
- [x] Update the CHANGELOG.md
- [x] Check if full simulation tests pass locally (`EXECUTE_TESTS_ON=master pytest`)

Also the following steps were realized (if applies):
- [x] Use in-line comments to explain your code
- [x] Write docstrings to your code
- [ ] For new functionalities: Explain in readthedocs
- [ ] Write test(s) for your new patch of code

*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

In case of an error due to linting, run `black . --exclude docs/` and push your changes.
Note that in case you do not fix a whole issue you should start your PR with `Address #xyz`.

<sub>*For more information on how to contribute check the [CONTRIBUTING.md](https://github.com/rl-institut/mvs_eland/blob/dev/CONTRIBUTING.md).*<sub>

### CODE REVIEW

File: tests/test_examples.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            constants.EXAMPLE_DIRECTORY, "run_pvcompare_example_electricity_sector.py"
+        )
+        self.coupled_sector_path = os.path.join(
+            constants.EXAMPLE_DIRECTORY, "run_pvcompare_example_sector_coupling.py"
+        )
+        self.coupled_sector_gas_path = os.path.join(
+            constants.EXAMPLE_DIRECTORY, "run_pvcompare_example_sector_coupling_gas.py"
+        )
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
```
Comment (SabineHaas) [inline tests/test_examples.py:37]:
```suggestion
    # this ensures that the test is only run if explicitly executed, i.e. not when the `pytest` command
```

File: tests/test_examples.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    # # alone is called
```
Comment (SabineHaas) [inline tests/test_examples.py:38]:
```suggestion
    # alone is called
```

File: tests/test_examples.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    )
+    @mock.patch("argparse.ArgumentParser.parse_args", return_value=argparse.Namespace())
+    def test_run_pvcompare_example_electricity_sector(self, margs):
+        exit_code = 1
+        scenario_name = "Scenario_example_electricity_sector"
+        # Read run_pvcompare_example_electricity_sector.py
+        elec_sector = open(self.elec_sector_path).read()
+
+        # Modify the name of the scenario to ensure it will run through for the test
+        # if the user already has it in the examples output folder
```
Comment (SabineHaas) [inline tests/test_examples.py:52]:
```suggestion
        # also if the user already has it in the examples output folder
```

File: tests/test_examples.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        )
+        self.coupled_sector_gas_path = os.path.join(
+            constants.EXAMPLE_DIRECTORY, "run_pvcompare_example_sector_coupling_gas.py"
+        )
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
+    # # alone is called
+    @pytest.mark.skipif(
+        EXECUTE_TESTS_ON not in (TESTS_ON_MASTER),
+        reason="Benchmark test deactivated, set env variable "
```
Comment (SabineHaas) [inline tests/test_examples.py:41]:
```suggestion
        reason="Test deactivated, set env variable "
```

File: tests/test_examples.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        elec_sector_modified = open(self.elec_sector_path, "w")
+        elec_sector_modified.write(elec_sector)
+        elec_sector_modified.close()
+
+        # Delete example output of tests
+        dir_name = os.path.join(
+            self.outputs_directory, scenario_name + "_test_run_through",
+        )
+        if os.path.exists(dir_name):
+            shutil.rmtree(dir_name, ignore_errors=True)
```
Comment (SabineHaas) [inline tests/test_examples.py:78]: This part (deleting files) could be a teardown method or teardown class to avoid doubling.

**Review by SabineHaas — COMMENTED** [review]
Thank you @MaGering for adding these tests!

As you have said yourself there are comments that need adaption.

Where do you actually run the test? Does `if os.system("python " + self.elec_sector_path) == 0:` run the python file?

File: tests/test_examples.py (hunk 5 above)
Comment (MaGering) [reply tests/test_examples.py]: Done with commit https://github.com/greco-project/pvcompare/pull/284/commits/68b087bb4fb9fc5866ca24edcb8a5b02da663e41.

**Comment by MaGering** [comment]
> Where do you actually run the test? Does `if os.system("python " + self.elec_sector_path) == 0:` run the python file?

Yes exactly, it executes the script and returns the exit code in 16 bits code. I explained this (with recent commit https://github.com/greco-project/pvcompare/pull/284/commits/3cee62401d7c6a7399aafb3f24bc2fb68ee9dc05) in a comment in the lines above `if os.system("python " + self.elec_sector_path) == 0:` .
But basically `exit code 0` leads to `0` and `exit code 1` leads to `256`.

**Comment by MaGering** [comment]
@SabineHaas do you know why the Read the Docs build fails? :-/

**Comment by SabineHaas** [comment]
> @SabineHaas do you know why the Read the Docs build fails? :-/

Seems like there's a problem with the installation of `mock`.. However, for #275 it works..
You could try to merge #275 to dev and merge changes into this PR.
If it still fails then we need to check further..

File: tests/test_examples.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+        # Revert changes made in the example file
+        elec_sector_modified = open(self.elec_sector_path, "w")
+        elec_sector_modified.write(elec_sector)
+        elec_sector_modified.close()
+
+        # Delete example output directory of this test
+        TestExamples.teardown_method(self)
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
```
Comment (SabineHaas) [inline tests/test_examples.py:102]:
```suggestion
    # this ensures that the test is only run if explicitly executed, i.e. not when the `pytest` command
```

File: tests/test_examples.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        # Revert changes made in the example file
+        coupled_sector_modified = open(self.coupled_sector_path, "w")
+        coupled_sector_modified.write(coupled_sector)
+        coupled_sector_modified.close()
+
+        # Delete example output directory of this test
+        TestExamples.teardown_method(self)
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
```
Comment (SabineHaas) [inline tests/test_examples.py:146]:
```suggestion
    # this ensures that the test is only run if explicitly executed, i.e. not when the `pytest` command
```

File: tests/test_examples.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        # Revert changes made in the example file
+        elec_sector_modified = open(self.elec_sector_path, "w")
+        elec_sector_modified.write(elec_sector)
+        elec_sector_modified.close()
+
+        # Delete example output directory of this test
+        TestExamples.teardown_method(self)
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
+    # # alone is called
```
Comment (SabineHaas) [inline tests/test_examples.py:103]:
```suggestion
    # alone is called
```

File: tests/test_examples.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        # Revert changes made in the example file
+        coupled_sector_modified = open(self.coupled_sector_path, "w")
+        coupled_sector_modified.write(coupled_sector)
+        coupled_sector_modified.close()
+
+        # Delete example output directory of this test
+        TestExamples.teardown_method(self)
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
+    # # alone is called
```
Comment (SabineHaas) [inline tests/test_examples.py:147]:
```suggestion
    # alone is called
```

File: tests/test_examples.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        elec_sector_modified.close()
+
+        # Delete example output directory of this test
+        TestExamples.teardown_method(self)
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
+    # # alone is called
+    @pytest.mark.skipif(
+        EXECUTE_TESTS_ON not in (TESTS_ON_MASTER),
+        reason="Benchmark test deactivated, set env variable "
```
Comment (SabineHaas) [inline tests/test_examples.py:106]:
```suggestion
        reason="Test deactivated, set env variable "
```

File: tests/test_examples.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        coupled_sector_modified.close()
+
+        # Delete example output directory of this test
+        TestExamples.teardown_method(self)
+
+    # # this ensure that the test is only ran if explicitly executed, ie not when the `pytest` command
+    # # alone is called
+    @pytest.mark.skipif(
+        EXECUTE_TESTS_ON not in (TESTS_ON_MASTER),
+        reason="Benchmark test deactivated, set env variable "
```
Comment (SabineHaas) [inline tests/test_examples.py:150]:
```suggestion
        reason="Test deactivated, set env variable "
```

**Review by SabineHaas — APPROVED** [review]
Could you add an assertion statement behind the assertion e.g.
`, f"The example 'xy' exited with exit code {exit_code}." ?

Then everything looks good from my point of view.

**Comment by MaGering** [comment]
> Could you add an assertion statement behind the assertion e.g.
> `, f"The example 'xy' exited with exit code {exit_code}." ?

Good point! Done with commit https://github.com/greco-project/pvcompare/pull/284/commits/5952083ac4d83f97a9b462e9e9db0f67cb6238f5.

**Pull request merged**